### PR TITLE
Ip/nac 1514. Mechanism to re-queue messages in poison queue

### PR DESCRIPTION
**Pull request by ipiatrou**
## What’s changing?

Added azure function (PoisonMessageDequeuer) to re-queue messages in the poison queue.

## Why?

NAC-1514

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [x] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [x] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [x] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [x] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -23,3 +23,7 @@ iac/env/fns/**
 liquibase.properties
 .runsettings
 coveragereport/
+
+support-tools/src/Tests/Piipan.SupportTools.Func.Api.Tests/bin/
```
Comment (bgfoshay) [inline .gitignore:27]: Question: Are these exclusions necessary? Just curious because we don't explicitly list any other test project folders

File: iac/create-resources.bash [hunk 2]
```diff
@@ -604,6 +604,82 @@ main () {
...
+      location="$LOCATION" \
+      functionAppName="$SUPPORT_TOOLS_FUNC_APP_NAME" \
+      appServicePlanName="$APP_SERVICE_PLAN_FUNC_NAME" \
+      storageAccountName="$SUPPORT_TOOLS_FUNC_APP_STORAGE_NAME" \
+      collabDatabaseConnectionString="" \
+      cloudName="$CLOUD_NAME" \
+      coreResourceGroup="$RESOURCE_GROUP" \
+      diagnosticSettingName="${DIAGNOSTIC_SETTINGS_NAME}" \
+      eventHubAuthorizationRuleId="${EH_RULE_ID}" \
+      eventHubName="" \
```
Comment (bgfoshay) [inline iac/create-resources.bash:625]: Question: Should ${EVENT_HUB_NAME} be provided here so that Event Hub is configured for the Function?

File: iac/create-resources.bash [hunk 3]
```diff
@@ -604,6 +604,82 @@ main () {
...
+    --vnet "$VNET_ID"
+
+  echo "Removing public access for ${SUPPORT_TOOLS_FUNC_APP_NAME}"
+  az storage account update \
+    --name "${SUPPORT_TOOLS_FUNC_APP_STORAGE_NAME}" \
+    --resource-group "${RESOURCE_GROUP}" \
+    --default-action "Deny"
+
+  # Update Key Vault to allow function access
+  echo "Granting Key Vault access for ${SUPPORT_TOOLS_FUNC_APP_NAME}"
```
Comment (bgfoshay) [inline iac/create-resources.bash:648]: Question: Considering this function doesn't access the databases, do we need to give the function access to Key Vault?  In the interest of least privilege, we may be able to skip granting this function access.

File: support-tools/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+3. Add the request body as shown in the provided JSON format 
+```JSON
+{
+ "queue_name": "{YOUR QUEUE NAME}",
+ "account_name": "{YOUR ACCOUNT NAME}",
+ "account_key": "{YOUR ACCOUNT KEY}"
+}
+```
+4. Send the POST HTTP request to the specified URL
+`https://{RERFIX}-func-support-tools-dev.azurewebsites.net/api/PoisonMessageDequeuer`
```
Comment (bgfoshay) [inline support-tools/README.md:37]: Nitpick: typo. I think "RERFIX" should be "PREFIX"? Also, rather than specify "dev", we may want to list {env} as an environment placeholder. e.g. `https://{PREFIX}-func-support-tools-{ENV}.azurewebsites.net/api/PoisonMessageDequeuer`

File: support-tools/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (bgfoshay) [inline support-tools/README.md:38]: Question: Are there any environment variables that need to be called out in a section for Environment Variables?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                // have not been deleted will become visible again.
+
+                QueueMessage[] messages = await poisonQueue.ReceiveMessagesAsync(20, TimeSpan.FromMinutes(10));
+
+                if (messages == null || messages.Length == 0)
+                    break;
+
+                foreach (var message in messages)
+                {
+                    await poisonQueue.DeleteMessageAsync(message?.MessageId, message?.PopReceipt);
```
Comment (dborgen2) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:49]: It looks like both DeleteMessageAsync and SendMessageAsync return some values. I think we should probably double check the delete was successful before sending a new one, and then if the send is successful we can then increment the count. Also, if for some reason DeleteMessageAsync fails we will be stuck in the while loop, so we may want to consider breaking out of the upper while loop if any of the messages fail to delete (or add in some retry mechanism like Polly).

File: .gitignore (hunk 1 above)
Comment (ipiatrou) [reply .gitignore:27]: Yep. It isn't necessary. I will remove it.

File: iac/configure-easy-auth.bash [hunk 7]
```diff
@@ -273,6 +276,7 @@ main () {
   NOTIFY_API_APP_ROLE='Notify.Read'
   ORCH_API_APP_ROLE='OrchestratorApi.Query'
   STATES_API_APP_ROLE='States.Query'
+  SUPPORT_TOOLS_API_APP_ROLE='SupportTools'
```
[the remark this answers is not in the document]
Comment (ipiatrou) [reply iac/configure-easy-auth.bash:279]: Based on my understanding, this role "States.Query" has this name because StatesApi has a connection to the Query tool.

File: support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                }
+
+                var storageAccountString = $"DefaultEndpointsProtocol=https;AccountName={param.AccountName};AccountKey={param.AccountKey}";
+
+                var targetQueue = new QueueClient(storageAccountString, param.QueueName);
+                var poisonQueue = new QueueClient(storageAccountString, param.QueueName + "-poison");
+
+                var countOfMessages = await _poisonMessageService.RetryPoisonMesssages(targetQueue, poisonQueue);
+
+                return new OkObjectResult($"Recived {countOfMessages} messages from {param.QueueName}-poison queue " +
```
Comment (dborgen2) [inline support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs:58]: Nitpick - Mis-spelled "Received"

File: support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            var two = new PoisonMessageDequeuerParam
+            {
+                AccountKey = one.AccountKey,
+                AccountName = one.AccountName,
+                QueueName = one.QueueName
+            };
+
+
+            // Act / Assert
+            Assert.True(one.QueueName.Equals(two.QueueName));
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs:27]: Food for thought (non-blocking, research needed): You may want to consider making PoisonMessageDequeuerParam a record or struct, since its intention is to be an immutable request. Then you could compare by value with just one Assert.Equal.

File: iac/create-resources.bash (hunk 2 above)
Comment (ipiatrou) [reply iac/create-resources.bash:625]: Right now Support tools do not use the event hub.
What is the role of Event Hubs in support tools?

File: iac/create-resources.bash (hunk 3 above)
Comment (ipiatrou) [reply iac/create-resources.bash:648]: That make sense I will skip it.

File: iac/create-resources.bash (hunk 2 above)
Comment (ryangeorgeharvey) [reply iac/create-resources.bash:625]: EventHub is where we send all of our logs (diagnostic settings in Azure), which then get sent to ISD Splunk. So yes, regardless of whether the function uses EventHub or not, it is still required to set those variables.

File: support-tools/tests/Piipan.SupportTools.Core.Tests/PoisonMessageServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            mockMessageReponse.SetupGet(m => m.Value).Returns(new QueueMessage[countOfMessages]);
+
+            var mockMessageReponseSecond = new Mock<Response<QueueMessage[]>>();
+            mockMessageReponseSecond.SetupGet(m => m.Value);
+
+            var mockExistResponse = new Mock<Response<bool>>();
+            mockExistResponse.SetupGet(m => m.Value).Returns(true);
+
+            mockPoisonQueueClient.Setup(c => c.Exists(It.IsAny<CancellationToken>())).Returns(mockExistResponse.Object);
+            mockPoisonQueueClient.SetupSequence(c => c.ReceiveMessagesAsync(It.IsAny<int>(), It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()))
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Core.Tests/PoisonMessageServiceTests.cs:36]: Minor nitpick (non-blocking) - I think to better test the while loop for large quantities of messages it would be better to test them 20 at a time like we would be doing in production. Feel free to use or not use, but I was thinking something like this:

```
var setupSequence = mockPoisonQueueClient.SetupSequence(c => c.ReceiveMessagesAsync(It.IsAny<int>(), It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()));
int sequencesAdded = 1; // always has at least one (the ending one)
for (int i = 0; i < countOfMessages; i += PoisonMessageService.MaxMessagesPerRequest)
{
    var mockMessageReponse = new Mock<Response<QueueMessage[]>>();
    if (i + PoisonMessageService.MaxMessagesPerRequest >= countOfMessages)
    {
        mockMessageReponse.SetupGet(m => m.Value).Returns(new QueueMessage[countOfMessages - i]);
    }
    else
    {
        mockMessageReponse.SetupGet(m => m.Value).Returns(new QueueMessage[PoisonMessageService.MaxMessagesPerRequest]);
    }

    setupSequence = setupSequence.Returns(Task.FromResult(mockMessageReponse.Object)); // not sure if assignment is needed here?
    sequencesAdded++;
}
setupSequence.Returns(Task.FromResult(mockEmptyResponse.Object));`

Then below you could verify the amount: `It.IsAny<CancellationToken>()), Times.Exactly(sequencesAdded));`

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        /// <summary>
+        /// Dequeues messages from poison queue and sends them to reqular queue
+        /// </summary>
+        /// <param name="storageAccountString"></param>
+        /// <param name="queueName"></param>
+        /// <returns></returns>
+        /// <exception cref="Exception"></exception>
+        public async Task<int> RetryPoisonMesssages(QueueClient targetQueue, QueueClient poisonQueue)
```
Comment (dborgen2) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:25]: Nitpick (non-blocking): Misspelling of "Messages" (extra s)

File: iac/create-resources.bash (hunk 3 above)
Comment (ryangeorgeharvey) [reply iac/create-resources.bash:648]: It needs Key Vault access in order to pull the Column Encryption key. However, looking at the function, I don't think it connects to either DB. If the function doesn't need DB access, it doesn't need the appsetting for Column Encryption. And if it doesn't need that appsetting, it doesn't need Key Vault access.

File: iac/create-resources.bash (hunk 2 above)
Comment (ipiatrou) [reply iac/create-resources.bash:625]: Thanks for the explanation

File: iac/create-resources.bash (hunk 3 above)
Comment (ipiatrou) [reply iac/create-resources.bash:648]: Good to know.

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+using Moq;
+using Piipan.SupportTools.Core.Models;
+using Piipan.SupportTools.Core.Service;
+using Xunit;
+
+namespace Piipan.SupportTools.Func.Api.Tests
+{
+    public class PoisonMessageDequeuerTests
+    {
+        static Mock<HttpRequest> MockRequest()
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:15]: Question (non-blocking): Does this need to be static?

File: support-tools/README.md (hunk 4 above)
Comment (ipiatrou) [reply support-tools/README.md:37]: Actually, it isn't environment variables that we set up during creating Azure resources.

File: support-tools/README.md (hunk 5 above)
Comment (ipiatrou) [reply support-tools/README.md:38]: Actually, it isn't environment variables that we set up during creating Azure resources.

File: support-tools/README.md (hunk 4 above)
Comment (ipiatrou) [reply support-tools/README.md:37]: Fixed

File: iac/configure-easy-auth.bash (hunk 7 above)
[the remark this answers is not in the document]
Comment (ryangeorgeharvey) [reply iac/configure-easy-auth.bash:279]: I think this needs to be tested, but I don't think the role is needed at all. If the goal is just to enable Easy Auth, you could just use the enable_easy_auth() method. Then you wouldn't have to alter the configure_easy_auth_pair() method, and you wouldn't need to add a new role.

If the intent is that this function may be used by other functions in the future, then maybe the role is needed? @bgfoshay thoughts on how Support Tools will be used in the future?

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            Assert.Equal("Target and/or Poison Queues do not exist. The Poison Queue must be named '{target queue name}-poison'", result?.Value);
+        }
+
+        [Fact]
+        public async Task PoisonMessageDequeuer_ReturnsOkResponse()
+        {
+            // Arrange
+            var mockPoisonMessageService = new Mock<IPoisonMessageService>();
+            mockPoisonMessageService.Setup(c => c.Parse(It.IsAny<Stream>())).Returns(Task.FromResult(new PoisonMessageDequeuerParam()
+            {
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:114]: Minor nitpick (non-blocking): Can we break out some of this code into a construct function? It looks all the same except what mockPoisonMessageService does. Or maybe just have a DefaultPoisonMessageDequeuerParam variable.

Side note: I see AccountName is set to fakeQueueName, and QueueName is set to fakeAccountName. Is that on purpose? At first I thought the test was testing incorrect names but then I realized it was returning OK

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            var logger = new Mock<ILogger>();
+
+            // Act
+            var response = await api.Run(mockRequest.Object, logger.Object);
+
+            // Assert
+            var result = response as InternalServerErrorResult;
+            Assert.Equal(500, result.StatusCode);
+        }
+    }
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:152]: Should we add tests for if the Delete fails or if the Send fails? Meaning the exception is not thrown, but one of the two responses came back with a problem.

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/StartupTests.cs [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+        [Fact]
+        public void Configure_AllServicesResolve()
+        {
+            // Arrange
+            var tester = new DependencyTester()
+                .Register<PoisonMessageDequeuer>()
+                .Register<PoisonMessageService>();
+
+            // Act/Assert
+            tester.ValidateFunctionServices<Startup>();
```
Comment (dborgen2) [inline support-tools/tests/Piipan.SupportTools.Func.Api.Tests/StartupTests.cs:19]: Praise: Thanks for using the new DependencyTester!

**Review by ryangeorgeharvey — COMMENTED** [review]
Please update https://github.com/18F/piipan/blob/dev/docs/iac.md#environment-variables for the new function.

File: iac/configure-easy-auth.bash (hunk 7 above)
[the remark this answers is not in the document]
Comment (ipiatrou) [reply iac/configure-easy-auth.bash:279]: That is a good idea, I will test it.

File: support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs (hunk 8 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs:58]: Fixed

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 16]
```diff
@@ -0,0 +1,92 @@
+﻿using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Piipan.SupportTools.Core.Models;
+
+namespace Piipan.SupportTools.Core.Service
+{
+    public class PoisonMessageService : IPoisonMessageService
```
Comment (bgfoshay) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:9]: Nitpick: we need to document public facing classes, interfaces, & methods

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 11 above)
Comment (bgfoshay) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:25]: Nitpick: are the documented parameters correct? Looks like they differ from the actual parameters?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+            while (true)
+            {
+                // Receive and process 20 messages
+                // It also sets the invisibility timeout to five minutes for each message.
+                // Note that the five minutes starts for all messages at the same time, so after
+                // five minutes have passed since the call to ReceiveMessages, any messages which
+                // have not been deleted will become visible again.
+
+                QueueMessage[] messages = await poisonQueue.ReceiveMessagesAsync(20, TimeSpan.FromMinutes(10));
```
Comment (bgfoshay) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:42]: Question: The comments above say 5, but is the invisibility timeout 10 minutes?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 6 above)
Comment (bgfoshay) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:49]: I agree with Dan's comments. Additionally, should we be deleting from the poison queue prior to requeuing them in the target queue? i.e. only if we successfully requeue, then delete the message from the poison queue?

I also agree that we probably need a circuit breaker of some sort to avoid getting stuck in the while loop indefinitely.

File: support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.Extensions.Logging;
+using Piipan.SupportTools.Core.Service;
+
+namespace Piipan.SupportTools.Func.Api
+{
+    public class PoisonMessageDequeuer
```
Comment (bgfoshay) [inline support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs:15]: Nitpick: similar to earlier comments that we need class comments for public classes

File: support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+
+                var targetQueue = new QueueClient(storageAccountString, param.QueueName);
+                var poisonQueue = new QueueClient(storageAccountString, param.QueueName + "-poison");
+
+                var countOfMessages = await _poisonMessageService.RetryPoisonMesssages(targetQueue, poisonQueue);
+
+                return new OkObjectResult($"Recived {countOfMessages} messages from {param.QueueName}-poison queue " +
+                    $"and sent them to {param.QueueName} queue");
+            }
+            catch (HttpRequestException ex)
```
Comment (bgfoshay) [inline support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs:61]: Question: I don't think it's an issue, but I want to double-check: there's no way any PII or the AccountKey will be exposed in either of these exception methods, right?

File: support-tools/support tools.sln [hunk 20]
```diff
@@ -0,0 +1,49 @@
+﻿
```
Comment (bgfoshay) [inline support-tools/support tools.sln:1]: Question/Nitpick: Does this solution name have a space in the name? Can we instead have a dash? e.g. support-tools.sln?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 6 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:49]: Good catch.
I have fixed it.

File: iac/configure-easy-auth.bash (hunk 7 above)
[the remark this answers is not in the document]
Comment (ipiatrou) [reply iac/configure-easy-auth.bash:279]: I tested today. Unfortunately, it does not work.
When I use only the enable_easy_auth() method then I can't get-access-token.

[the remark this answers is not in the document]
Comment (ryangeorgeharvey) [reply iac/configure-easy-auth.bash:279]: Just to confirm, you only need access to the token for local development and debugging right? Azure wouldn't need it?

File: support-tools/tests/Piipan.SupportTools.Core.Tests/PoisonMessageServiceTests.cs (hunk 10 above)
Comment (ipiatrou) [reply support-tools/tests/Piipan.SupportTools.Core.Tests/PoisonMessageServiceTests.cs:36]: Yep, I did refactoring. For this purpose, I created the Batch method.

File: iac/configure-easy-auth.bash (hunk 7 above)
[the remark this answers is not in the document]
Comment (ryangeorgeharvey) [reply iac/configure-easy-auth.bash:279]: Still would make sense that you would need to connect for development, I just want to make sure.

File: support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs (hunk 9 above)
Comment (ipiatrou) [reply support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs:27]: Good idea. 
I am going to try next time record or struct.
Will it work for JsonConvert.DeserializeObject ?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 11 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:25]: fixed

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs (hunk 12 above)
Comment (ipiatrou) [reply support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:15]: Fixed

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs (hunk 13 above)
Comment (ipiatrou) [reply support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:114]: Make sense.

File: support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs (hunk 14 above)
Comment (ipiatrou) [reply support-tools/tests/Piipan.SupportTools.Func.Api.Tests/PoisonMessageDequeuerTests.cs:152]: Already added.

File: support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs (hunk 19 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Func.Api/PoisonMessageDequeuer.cs:61]: No, checked it.

File: support-tools/support tools.sln (hunk 20 above)
Comment (ipiatrou) [reply support-tools/support tools.sln:1]: No, a C# solution file name should not contain spaces. It is a common convention to use underscores or camelCase to separate words in the file name, but spaces are generally not used as they can cause issues with file system compatibility or accessing the file through the command line.
Fixed it

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 17 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:54]: Fixed

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 16 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:12]: Done

File: support-tools/src/Piipan.SupportTools.Core/Models/PoisonMessageDequeuerParam.cs [hunk 21]
```diff
@@ -0,0 +1,14 @@
+﻿using Newtonsoft.Json;
+
+namespace Piipan.SupportTools.Core.Models
+{
+    public class PoisonMessageDequeuerParam
```
[the remark this answers is not in the document]
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Models/PoisonMessageDequeuerParam.cs:8]: Fixed

File: support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs (hunk 9 above)
Comment (dborgen2) [reply support-tools/tests/Piipan.SupportTools.Core.Tests/Models/PoisonMessageDequeuerParamTests.cs:27]: Yep, it should, but you'd want to test to make sure. Under the hood records are simply classes with a few methods overloaded so definitely no problem there. Structs on the other hand I could see a problem since parameterless constructors were not available for structs until C# 10 (.Net 6), but since we are using that it should be fine.

I normally default to classes, but if I find myself comparing all of the values of the object then I consider one of these options.

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+                    if (sendMesResponse?.Value == null)
+                    {
+                        _logger.LogError($"RetryPoisonMesssages: SendMessageAsync:: problem with sending the message. Message Id = {message?.MessageId}");
+                        continue;
+                    }
+
+                    countOfMessages++;
+
+                    var delMesResponse = await poisonQueue.DeleteMessageAsync(message?.MessageId, message?.PopReceipt);
+                    if (delMesResponse?.IsError == true)
```
Comment (dborgen2) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:73]: Minor nitpick: Can delMesResponse actually be a null value? If so, we should probably log an error if it is null, correct?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+            //to avoid getting stuck in the while loop indefinitely
+            while (cachedMessagesCountInPoisonQueue > 0)
+            {
+                // Receive and process 20 messages
+                // It also sets the invisibility timeout to ten minutes for each message.
+                // Note that the five minutes starts for all messages at the same time, so after
+                // five minutes have passed since the call to ReceiveMessages, any messages which
+                // have not been deleted will become visible again.
+
+                QueueMessage[] messages = await poisonQueue.ReceiveMessagesAsync(20, TimeSpan.FromMinutes(10));
```
Comment (dborgen2) [inline support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:54]: Minor nitpick: This 20 might be better as a constant, so that if it changes for whatever reason we don't need to change our tests (assuming we update the tests to also use this constant)

**Comment by dborgen2** [comment]
LGTM, just a couple of minor things upon re-review. I know Bill has some open items so I won't approve it yet, but I think I'm good

File: iac/configure-easy-auth.bash (hunk 7 above)
[the remark this answers is not in the document]
Comment (ipiatrou) [reply iac/configure-easy-auth.bash]: I guess this azure function should be deployed to every environment (dev, pre-prod, prod)

1. The azure function has to have access to queues in Storage Accounts for the current env. This function (PoisonMessageDequeuer) is HTTP and we send a post request with the body 
{
 "queue_name": "{YOUR QUEUE NAME FROM STORAGE ACCOUNT}",
 "account_name": "{YOUR STORAGE ACCOUNT NAME}",
 "account_key": "{YOUR STORAGE ACCOUNT PRIMARY KEY}"
}
2. To execute the azure function we have to get an access token. Example:
az account get-access-token --resource "api://blablablablablablalbalbalba" --query accessToken -o tsv 

@bgfoshay 
Please correct me if I am wrong.

[the remark this answers is not in the document]
Comment (bgfoshay) [reply iac/configure-easy-auth.bash]: That's what I was thinking

File: iac/create-resources.bash [hunk 24]
```diff
@@ -604,6 +604,64 @@ main () {
...
+    --name "${SUPPORT_TOOLS_FUNC_APP_STORAGE_NAME}" \
+    --resource-group "${RESOURCE_GROUP}" \
+    --default-action "Deny"
+
+  echo "Update ${SUPPORT_TOOLS_FUNC_APP_NAME} settings"
+  az functionapp config appsettings set \
+    --name "${SUPPORT_TOOLS_FUNC_APP_NAME}" \
+    --resource-group "$RESOURCE_GROUP" \
+    --settings \
+      ${COLUMN_ENCRYPT_KEY}="@Microsoft.KeyVault(VaultName=${VAULT_NAME};SecretName=${COLUMN_ENCRYPT_KEY_KV})" \
```
Comment (bgfoshay) [inline iac/create-resources.bash:652]: Question: Does this function need the column encryption key?

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 17 above)
Comment (bgfoshay) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:54]: I think there are a still a couple references to "five" in the comments

**Review by bgfoshay — APPROVED** [review]
Praise: Thanks for all of the hard work on this one! I know we threw a lot of comments at you and asked for a lot of tweaks. I appreciate the patience and your working through all of them. This will be a useful utility for the support team.

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 17 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:54]: sorry. fixed

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 22 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:73]: corrected

File: support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs (hunk 23 above)
Comment (ipiatrou) [reply support-tools/src/Piipan.SupportTools.Core/Service/PoisonMessageService.cs:54]: sounds good.
Added 2 constants / updated unit tests

File: iac/create-resources.bash (hunk 24 above)
Comment (ipiatrou) [reply iac/create-resources.bash]: Removed. 
Tested.
It looks good
